# Tugas Event #2 - Solving Real-World Problems with Data Science

- Nama           : Nadif Aulia Putra
- NIM            : 16520318
- Universitas    : Institut Teknologi Bandung
- Tanggal Mulai  : 15 November 2020
- Sumber Dataset : https://www.kaggle.com/dansbecker/melbourne-housing-snapshot

(Sebenarnya data yang disediakan cukup banyak. Namun, untuk mempermudah analisis, maka akan digunakan data dari train.csv saja)

## 1. Data Preparation & Observation

- Kita akan melihat data tentang harga rumah.

- Disini aku menggunakan perpustakaan Pandas untuk projek ini. Pandas adalah alat utama yang digunakan para ilmuwan untuk menjelajahi dan memanipulasi data.

In [5]:
import pandas as pd

data_rumah = pd.read_csv("train.csv")

Di bawah ini kita akan melihat:
- count: menunjukkan berapa banyak baris yang memiliki nilai yang tidak hilang
- mean: rata-rata
- std: standar deviasi
- nilai min, 25%, 50%, 75%, dan maks

In [6]:
data_rumah.describe()

Id   MSSubClass  ...       YrSold      SalePrice
count  1460.000000  1460.000000  ...  1460.000000    1460.000000
mean    730.500000    56.897260  ...  2007.815753  180921.195890
std     421.610009    42.300571  ...     1.328095   79442.502883
min       1.000000    20.000000  ...  2006.000000   34900.000000
25%     365.750000    20.000000  ...  2007.000000  129975.000000
50%     730.500000    50.000000  ...  2008.000000  163000.000000
75%    1095.250000    70.000000  ...  2009.000000  214000.000000
max    1460.000000   190.000000  ...  2010.000000  755000.000000

[8 rows x 38 columns]

Dari program diatas kita dapat mengetahui rumah terbaru yang ada di dalam data ini yang dapat diliha dari nilai max dalam kolom "YrSold"

Data rumah setelah 2010 tidak ditemukan sehingga kemungkinannya ada 2:
1. Mereka belum membangun rumah baru di tempat pengumpulan data ini.
2. Data dikumpulkan sejak lama. Rumah yang dibangun setelah publikasi data tidak diperbarui ke dalam data

Ada beberapa hal yang dapat disimpulkan:
- Dataset mungkin sudah usang.

- Hal ini dapat diverifikasi dengan membandingkan data perumahan ini dengan yang tercantum dalam kumpulan data, dan melihat tahun terakhir pembuatan, tahun penjualan, dll. Jika salah satu dari data ini melebihi tahun 2011, dapat dipastikan bahwa kumpulan data tersebut sudah usang.

- Ini menjadi masalah karena data ini tidak relevan dan saya hanya menggunakannya untuk mengeksplorasi saja.

## 2. Model Development

- Disini saya menggunakan notasi titik untuk memilih kolom yang ingin diprediksi, yang disebut target prediksi.
- Secara konvensi, target prediksi disebut y.

In [7]:
y = data_rumah.SalePrice

- Kolom yang dimasukkan ke dalam model (dan kemudian digunakan untuk membuat prediksi) disebut *features*
- Ini adalah kolom yang digunakan untuk menentukan harga rumah
- Secara konvensi disebut X

In [8]:
features = ['LotArea', 'YearBuilt', '1stFlrSF', '2ndFlrSF', 'FullBath', 'BedroomAbvGr', 'TotRmsAbvGrd']
X = data_rumah[features]

- Library scikit-learn mempunyai fungsi train_test_split untuk memecah data menjadi dua bagian

- Saya menggunakan beberapa dari data itu sebagai data pelatihan agar sesuai dengan model, dan kami akan menggunakan data lain sebagai data validasi untuk menghitung mean_absolute_error

In [9]:
from sklearn.model_selection import train_test_split

train_X, val_X, train_y, val_y = train_test_split(X, y, random_state=1)


- Membuat model menggunakan *random forest*

- *Random forest* menggunakan banyak *trees*, dan membuat prediksi dengan merata-ratakan prediksi setiap pohon komponen.

- Biasanya memiliki akurasi prediksi yang jauh lebih baik daripada *single decision tree*

In [10]:
from sklearn.ensemble import RandomForestRegressor

rf_model = RandomForestRegressor(random_state=1)

- Fit: Menemukan pola dari data yang disediakan

In [11]:
rf_model.fit(train_X, train_y)

RandomForestRegressor(bootstrap=True, ccp_alpha=0.0, criterion='mse',
                      max_depth=None, max_features='auto', max_leaf_nodes=None,
                      max_samples=None, min_impurity_decrease=0.0,
                      min_impurity_split=None, min_samples_leaf=1,
                      min_samples_split=2, min_weight_fraction_leaf=0.0,
                      n_estimators=100, n_jobs=None, oob_score=False,
                      random_state=1, verbose=0, warm_start=False)

- Membuat prediksi dengan perintah prediksi model menggunakan val_X sebagai datanya

- Simpan hasilnya ke variabel yang disebut rf_val_predictions

In [12]:
rf_val_predictions = rf_model.predict(val_X)

- Kesalahan absolut rata-rata / *mean absolute error* (MAE) adalah error dengan melihat perbedaan mutlak antara nilai aktual dan perkiraan lalu dicari rata-ratanya

In [ ]:
from sklearn.metrics import mean_absolute_error

rf_val_mae = mean_absolute_error(rf_val_predictions, val_y)

print("Validation MAE for Random Forest Model: {}".format(rf_val_mae))

- Metode persentase kesalahan rata-rata absolut / *mean absolute persentage error* (MAPE) adalah metode persentase error dibandingkan dengan nilai sebenarnya

- Ini memberikan lebih banyak ukuran kesalahan standar. Misalnya, jika kesalahannya adalah 10 dan nilai sebenarnya adalah 100, maka persentasenya akan menjadi 10% dibandingkan dengan jika kesalahannya 100 dan nilai sebenarnya adalah 1000, ukurannya tetap 10%

In [ ]:
import numpy as np

def mean_absolute_percentage_error(y_true, y_pred):
    y_true, y_pred = np.array(y_true), np.array(y_pred)
    return np.mean(np.abs((y_true - y_pred) / y_true)) * 100

mean_absolute_percentage_error(val_y, rf_val_predictions)

14.12783863744737

## 3. Presentation

- Disini saya membuat model untuk menentukan harga rumah berdasarkan data yang ada

- Dapat dilihat bahwa mean absolute persentage error (MAPE) dari model machine learning yang dibuat adalah 14.13%, yang berarti keakuratan model yang dibuat mencapai 85.87%

- Menurut saya angka ini sudah cukup baik dan mungkin masih dapat ditingkatkan keakuratannya dengan model yang lain